### ISSUE TITLE
Git/Github documentation enhancements

### ISSUE THREAD

**Issue by gwhitney** [numberscope/frontscope#221]
We are seeing a recurring phenomenon where PRs become very convoluted, apparently because of various `git merge` operations being used and replaying some actions. As one way to address this situation, I have some recommendations about the docsite with respect to Git/GitHub/code maintenance.

1) Nowhere do we suggest that contributors not use `git merge` for maintaining their branches; we should specifically mention this somewhere to amplify/expand on the recommendation that they do use `git rebase`.

2) Similarly, we need to make it clear in the appropriate place(s) never to commit to main, even in your fork.

3) The docsite currently has quite a few pages that go over things like code setup and git/github operations: https://numberscope.colorado.edu/doc/doc/running-from-source/, https://numberscope.colorado.edu/doc/CONTRIBUTING/, https://numberscope.colorado.edu/doc/doc/onboarding/,  https://numberscope.colorado.edu/doc/doc/working-with-git-and-github/, https://numberscope.colorado.edu/doc/doc/gitting-it-right/, https://numberscope.colorado.edu/doc/doc/pull-request-checklist/. I think this ends up being a bit confusing/hard to follow. Ideally, we would determine the information that needs to be covered, and segment it clearly into some of these pages, and try to avoid duplicating material in different pages.

4) We should pick one scheme of cloning/remote naming and stick to it everywhere. Currently most things are "clone from fork, so that the fork is origin, and add the official repo as upstream". That organization definitely works, so we could make those instructions explicit and consistent everywhere (they currently are not). I actually use "clone from the official repo, so that the official is origin, and add the fork as fork". I personally find the latter easier to keep track of and less error-prone, but I don't want to enforce my perspective on everyone. We should discuss and pick a scheme, and make sure all the docs are consistent with it. If the consensus is "fork = origin, official = upstream" then I will switch to that scheme so that I am dogfooding our instructions.

3) Worse, there are a number of places where one page contradicts another. The remaining points in this list catalog a number of such instances. I am not making any claim that this list is exhaustive.

4) https://numberscope.colorado.edu/doc/CONTRIBUTING/ says to sync a fork with the Github UI. https://numberscope.colorado.edu/doc/doc/working-with-git-and-github/ has a different command-line procedure, that omits the step of pushing back to the fork (and doesn't happen to be the one I used before the "sync fork" button on the UI. We should pick one and just recommend that. If we emphasize no commits to main, then the Github UI is the simplest way to go.

5) On https://numberscope.colorado.edu/doc/doc/running-from-source/ we recommend VScode. On https://numberscope.colorado.edu/doc/doc/onboarding/ we say there isn't an editor we recommend. And then on https://numberscope.colorado.edu/doc/doc/visual-studio-code-setup/ we recommend VScode + Volar. We can offer info about VScode, but we should be consistent in how we frame it: as info-only or as a recommendation. I prefer the former because I's prefer not to use VScode, but if we make it an official recommendation I will try to dogfood it.

OK, I am sure this is not yet exhaustive but I have typed enough. I will add more internal contradictions as time permits.

**Comment by gwhitney** [comment]
Amplifying item 4 from above, https://numberscope.colorado.edu/doc/doc/running-from-source/ says to clone the official repo, https://numberscope.colorado.edu/doc/CONTRIBUTING/ says to clone a fork, https://numberscope.colorado.edu/doc/doc/onboarding/ prefers to clone a fork but mentions you can clone the official repo if you don't want to bother forking, and the "Push a branch" section of https://numberscope.colorado.edu/doc/doc/working-with-git-and-github/ strongly implies that you cloned a fork.

**Comment by gwhitney** [comment]
Item 10 in https://numberscope.colorado.edu/doc/CONTRIBUTING/ links to a non-existent heading in https://numberscope.colorado.edu/doc/doc/working-with-git-and-github/. But then it has its own "Submit a pull request" section. That section mentions to sync your fork, but the pull request checklist does not. And neither place ever mentions to be sure the branch you are submitting as the PR is based on current main, as it should be. Suggestion: move the steps for actually submitting the pull request into the pull request checklist, and just link there from everywhere that talks about submitting a request, and get all of the steps straight in that one location.

**Comment by gwhitney** [comment]
Not a contradiction, but a suggestion: in the push changes section of https://numberscope.colorado.edu/doc/doc/working-with-git-and-github/ it just says to `git push`. But a newly-created local branch on the user's machine would not generally be set up to track any remote, so the `git push` would not know what to do. So there should be some further explanation about how/when to do something like `git push myremote mybranch` and how to set up tracking if desired (I don't think just pushing once does it, but am not really sure as I never bother to set up tracking myself, I just always use the full form with a remote and a branch).

**Comment by gwhitney** [comment]
Another suggestion: the stashing section on that same working with git and github page says

> Alternatively, if you wish to discard those changes from the stash, issue the following command:

I think it should make clear that the `pop` command both applies the changes and subsequently discards them from the stash, not just discards them (I think there is a different command that does that...)

**Comment by gwhitney** [comment]
The "Create a Branch" section on that same page says:

> To create a branch and switch to it in one command, issue the following command:

Doesn't it make a difference what branch you currently have checked out when you do this? Shouldn't it tell you to first check out the branch that you want the new branch to be created from? (often `main`?)

**Comment by gwhitney** [comment]
The "Push Changes" and "Push a Branch" sections on that page are confusing and potentially seem contradictory. There should either be just one section or they should be carefully made complementary and refer to each other.

**Comment by liammulh** [comment]
> I think it should make clear that the pop command both applies the changes and subsequently discards them from the stash, not just discards them (I think there is a different command that does that...)

The command you are thinking of is `git stash apply`, I think.

**Comment by gwhitney** [comment]
> > I think it should make clear that the pop command both applies the changes and subsequently discards them from the stash, not just discards them (I think there is a different command that does that...)
> 
> The command you are thinking of is `git stash apply`, I think.

No, `git stash apply` is already mentioned in our docs. That applies but does not discard. `git stash pop` applies and discards. I think there is yet a third command that discards without applying.

**Comment by liammulh** [comment]
In today's meeting we discussed what we think our Git workflow should be:

1. Clone the original numberscope/frontscope repo.
2. Create a branch in that repo.
3. Once you have a unit that can be considered for a PR, create a fork.
4. Push your branch to the fork.
5. Open a PR based on the branch in your fork.

This makes updating the main branch easier than if you were to fork then create a branch in your fork.

**Comment by gwhitney** [comment]
Note that recent prolific contributor @Vectornaut prefers merge to rebase; there should be a project discussion with him to understand the pros and cons before finalizing what to recommend in our documentation.

**Comment by gwhitney** [comment]
I think the delft PR series demonstrates the need to use rebase, so I would consider the strategy here settled. This should be part of the documentation overhaul for alpha.

**Comment by gwhitney** [comment]
Resolved in ui2 per #463. Closing.
